substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+  // 1) The coins are small, easy to run, and make no sense to be in their own set
+  // In this case, it's still hard to ask validators to run 16 different nodes
+  // 2) The coins are all on the same network yet there's no DEX on-chain
+  // In these cases, it'd be hard to find and justify 16 different coins from that single chain
+  // This could probably be just 8, yet 16 is a hedge for the unforseen
+  // If necessary, this can be increased with a fork
+  type MaxCoinsPerSet = ConstU32<16>;
+
+  // Support keys up to 96 bytes (BLS12-381 G2)
+  const MAX_KEY_LEN: u32 = 96;
+  type MaxKeyLen = ConstU32<MAX_KEY_LEN>;
+
+  #[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+  pub struct ValidatorSet<T: Config> {
+    bond: Amount,
+    coins: BoundedVec<Coin, MaxCoinsPerSet>,
+
+    // Participant and their amount bonded to this set
+    // Limit each set to 100 participants for now
+    participants: BoundedVec<(T::AccountId, Amount), ConstU32<100>>,
+  }
+
+  #[pallet::pallet]
+  #[pallet::generate_store(pub(super) trait Store)]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  /// The details of a validator set instance.
+  #[pallet::storage]
+  #[pallet::getter(fn validator_set)]
+  pub type ValidatorSets<T: Config> =
+    StorageMap<_, Twox64Concat, ValidatorSetInstance, ValidatorSet<T>, OptionQuery>;
+
+  type Key = BoundedVec<u8, MaxKeyLen>;
+
+  /// The key for a given validator set instance coin.
+  #[pallet::storage]
+  #[pallet::getter(fn key)]
+  pub type Keys<T: Config> =
+    StorageMap<_, Twox64Concat, (ValidatorSetInstance, Coin), Key, OptionQuery>;
+
+  /// If an account has voted for a specific key or not. Prevents them from voting multiple times.
+  #[pallet::storage]
+  #[pallet::getter(fn voted)]
+  pub type Voted<T: Config> = StorageMap<_, Blake2_128Concat, (T::AccountId, Key), (), OptionQuery>;
+
+  /// How many times a key has been voted for. Once consensus is reached, the keys will be adopted.
+  #[pallet::storage]
+  #[pallet::getter(fn vote_count)]
+  pub type VoteCount<T: Config> =
+    StorageMap<_, Blake2_128Concat, (ValidatorSetInstance, Coin, Key), u16, ValueQuery>;
+
+  #[pallet::genesis_build]
+  impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+    fn build(&self) {
+      // Start at 1 to leave 0 available for a future design where Serai itself has a distinct
+      // validator set for its Tendermint consensus
+      let mut coins = Vec::new();
+      for coin in 0 .. self.coins.0 {
+        coins.push(Coin(coin));
+      }
+
+      let mut participants = Vec::new();
+      for participant in self.participants.clone() {
+        participants.push((participant, self.bond));
+      }
+
+      ValidatorSets::<T>::set(
+        ValidatorSetInstance(Session(0), ValidatorSetIndex(1)),
+        Some(ValidatorSet {
+          bond: self.bond,
+          coins: BoundedVec::try_from(coins).unwrap(),
+          participants: BoundedVec::try_from(participants).unwrap(),
```

Comment:
This is indexed by instance. It's intended to be immutable. Nothing here forces it to be though.

Reply:
Seems low-hanging and a good pattern to enforce immutability past this point yes? Once the set is compiled here it will not need to change unless a new session forms, yes?

Reply:
We can't force it to be immutable. All we can do is never write to it again. We can introduce niceties for that, such as a struct we'd have to explicitly bypass, but..

I will note we also may not want immutability. I'm unsure how removal due to slashes works under Parity's session pallet.